### PR TITLE
Nema phantom extension

### PR DESCRIPTION
**Pull request by kdulski**
Added:
- Modular way of creating phantom construction from materials such as water, aluminium, silica, plastic for which teh user can set a density in macro. 
- Various shapes (box, sphere, eliptical cylinder, cylinder) with rotation and translation added. 
-  More complex combination such as union subtraction and intersection of the phantom construction elements 
- Full positronium decay modes for nema points
- Implemented a dummy way to correlate density of the material with the positron reach and lifetime of direct positron-electron annihilation
- Three new histograms density of material in which annihilation occured as a function of X/Y/Z
- Spherical shape of nema point generation

Still can be done:
- general generation of the phantom construction material
- Proper definition of the positron range / direct lifetime as a function of density
- New histograms / fields in TTree : material ID / name in which annihilation took place

### CODE REVIEW

**Comment by JPETTomographyTester** [comment]
Found 9 violations:

**Reporter**: cppcheck-raporter
**Rule**: unusedFunction
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/examples/basic/basic.cc L51

The function &#39;Construct&#39; is never used.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a-37406236>*

**Reporter**: cppcheck-raporter
**Rule**: redundantAssignment
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/examples/external/external.cc L99

Variable &#39;cone_mesh&#39; is reassigned a value before the old one has been used if variable is no semaphore variable.. Variable &#39;cone_mesh&#39; is reassigned a value before the old one has been used. Make sure that this variable is not used like a semaphore in a threading environment before simplifying this code.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a1979725627>*

**Reporter**: cppcheck-raporter
**Rule**: redundantAssignment
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/examples/external/external.cc L93

Variable &#39;cone_mesh&#39; is reassigned a value before the old one has been used if variable is no semaphore variable.. Variable &#39;cone_mesh&#39; is reassigned a value before the old one has been used. Make sure that this variable is not used like a semaphore in a threading environment before simplifying this code.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a1541790971>*

**Reporter**: cppcheck-raporter
**Rule**: unusedFunction
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/src/BuiltInReader.cc L81

The function &#39;BuiltIn&#39; is never used.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a229782320>*

**Reporter**: cppcheck-raporter
**Rule**: noExplicitConstructor
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/tests/Simulator.hh L29

Class &#39;DetectorConstruction&#39; has a constructor with 1 argument that is not explicit. Such constructors should in general be explicit for type safety reasons. Using the explicit keyword in the constructor means some mistakes when using the class can be avoided.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a-2004644777>*

**Reporter**: cppcheck-raporter
**Rule**: unreadVariable
**Severity**: INFO
**File**: Core/DetectorConstruction.cpp L1369

Variable &#39;rotationNew&#39; is assigned a value that is never used.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a-1362421875>*

**Reporter**: cppcheck-raporter
**Rule**: variableScope
**Severity**: INFO
**File**: Core/DetectorConstruction.cpp L1331

The scope of the variable &#39;phantMatExt&#39; can be reduced. Warning: Be careful when fixing this message, especially when there are inner loops. Here is an example where cppcheck will write that the scope for &#39;i&#39; can be reduced:\012void f(int x)\012{\012    int i &#61; 0;\012    if (x) {\012        // it&#39;s safe to move &#39;int i &#61; 0;&#39; here\012        for (int n &#61; 0; n &lt; 10; ++n) {\012            // it is possible but not safe to move &#39;int i &#61; 0;&#39; here\012            do_something(&amp;i);\012        }\012    }\012}\012When you see this message it is always safe to reduce the variable scope 1 level.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a2017539277>*

**Reporter**: cppcheck-raporter
**Rule**: variableScope
**Severity**: INFO
**File**: Core/DetectorConstruction.cpp L1330

The scope of the variable &#39;phantomElementMaterial&#39; can be reduced. Warning: Be careful when fixing this message, especially when there are inner loops. Here is an example where cppcheck will write that the scope for &#39;i&#39; can be reduced:\012void f(int x)\012{\012    int i &#61; 0;\012    if (x) {\012        // it&#39;s safe to move &#39;int i &#61; 0;&#39; here\012        for (int n &#61; 0; n &lt; 10; ++n) {\012            // it is possible but not safe to move &#39;int i &#61; 0;&#39; here\012            do_something(&amp;i);\012        }\012    }\012}\012When you see this message it is always safe to reduce the variable scope 1 level.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a-1383315618>*

**Reporter**: cppcheck-raporter
**Rule**: unmatchedSuppression
**Severity**: INFO
**File**: Core/DetectorConstruction.cpp L70

Unmatched suppression: unusedFunction

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a1999427137>*

 *<ACCUMULATED-VIOLATIONS>*

**Comment by JPETTomographyTester** [comment]
Found 8 violations:

**Reporter**: cppcheck-raporter
**Rule**: unusedFunction
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/examples/basic/basic.cc L51

The function &#39;Construct&#39; is never used.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a-37406237>*

**Reporter**: cppcheck-raporter
**Rule**: redundantAssignment
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/examples/external/external.cc L99

Variable &#39;cone_mesh&#39; is reassigned a value before the old one has been used if variable is no semaphore variable.. Variable &#39;cone_mesh&#39; is reassigned a value before the old one has been used. Make sure that this variable is not used like a semaphore in a threading environment before simplifying this code.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a1979725627>*

**Reporter**: cppcheck-raporter
**Rule**: redundantAssignment
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/examples/external/external.cc L93

Variable &#39;cone_mesh&#39; is reassigned a value before the old one has been used if variable is no semaphore variable.. Variable &#39;cone_mesh&#39; is reassigned a value before the old one has been used. Make sure that this variable is not used like a semaphore in a threading environment before simplifying this code.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a1541790971>*

**Reporter**: cppcheck-raporter
**Rule**: unusedFunction
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/src/BuiltInReader.cc L81

The function &#39;BuiltIn&#39; is never used.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a229782319>*

**Reporter**: cppcheck-raporter
**Rule**: noExplicitConstructor
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/tests/Simulator.hh L29

Class &#39;DetectorConstruction&#39; has a constructor with 1 argument that is not explicit. Such constructors should in general be explicit for type safety reasons. Using the explicit keyword in the constructor means some mistakes when using the class can be avoided.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a-2004644777>*

**Reporter**: cppcheck-raporter
**Rule**: unreadVariable
**Severity**: INFO
**File**: Core/DetectorConstruction.cpp L1369

Variable &#39;rotationNew&#39; is assigned a value that is never used.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a-1362421876>*

**Reporter**: cppcheck-raporter
**Rule**: variableScope
**Severity**: INFO
**File**: Core/DetectorConstruction.cpp L1331

The scope of the variable &#39;phantMatExt&#39; can be reduced. Warning: Be careful when fixing this message, especially when there are inner loops. Here is an example where cppcheck will write that the scope for &#39;i&#39; can be reduced:\012void f(int x)\012{\012    int i &#61; 0;\012    if (x) {\012        // it&#39;s safe to move &#39;int i &#61; 0;&#39; here\012        for (int n &#61; 0; n &lt; 10; ++n) {\012            // it is possible but not safe to move &#39;int i &#61; 0;&#39; here\012            do_something(&amp;i);\012        }\012    }\012}\012When you see this message it is always safe to reduce the variable scope 1 level.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a2017539276>*

**Reporter**: cppcheck-raporter
**Rule**: unmatchedSuppression
**Severity**: INFO
**File**: Core/DetectorConstruction.cpp L70

Unmatched suppression: unusedFunction

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a1999427136>*

 *<ACCUMULATED-VIOLATIONS>*

File: Actions/PrimaryGeneratorAction.cpp [hunk 1]
```diff
@@ -187,33 +187,56 @@ void PrimaryGeneratorAction::SetNemaPositionWeight(G4int nemaPoint, G4int weight
...
     );
-  } else if (fNemaGenerator.DoesPointExistAlready(nemaPoint)) {
-    fNemaGenerator.SetPointLifetime(nemaPoint, lifetime);
-  } else {
+  } else if (!fNemaGenerator.DoesPointExistAlready(nemaPoint)) {
     fNemaGenerator.AddPoint(nemaPoint);
+  }
+  if (mode == PositroniumDecayMode::pPs) {
+    fNemaGenerator.SetPointPPsLifetime(nemaPoint, lifetime);
+  } else if (mode == PositroniumDecayMode::oPs) {
```
Comment (kkacprzak) [inline Actions/PrimaryGeneratorAction.cpp:202]: To keep it consistent, the function's name should be updated to `fNemaGenerator.SetPointOPsLifetime(nemaPoint, lifetime);`

File: Core/DetectorConstruction.h [hunk 2]
```diff
@@ -194,6 +220,9 @@ class DetectorConstruction : public G4VUserDetectorConstruction
   std::vector<Scin> fScinContainer;
   std::vector<Slot> fSlotContainer;
   G4int fLayerNumber = 0;
+
+  std::map<G4int, G4int> idPhantElem;
+  std::vector<PhantElem> phantomElements;
```
Comment (kkacprzak) [inline Core/DetectorConstruction.h]: I suggest following the convention and renaming these private members to `fPhantomElements` and `fPhantomElemIDs`

File: Core/DetectorConstruction.cpp [hunk 3]
```diff
@@ -1311,3 +1320,253 @@ void DetectorConstruction::CreateGeometryFile()
...
+      G4Material* phantomElementMaterial = nullptr;
+      G4Material* phantMatExt = nullptr;
+      G4LogicalVolume* phantomElementLogic = nullptr;
+      G4RotationMatrix* rotMat = new G4RotationMatrix();
+
+      if (phantomElements.at(i).fRotation.size() > 0) {
+        rotMat->rotateX(phantomElements.at(i).fRotation.at(0)*deg);
+        if (phantomElements.at(i).fRotation.size() > 1)
+          rotMat->rotateY(phantomElements.at(i).fRotation.at(1)*deg);
+        if (phantomElements.at(i).fRotation.size() > 2)
```
Comment (kkacprzak) [inline Core/DetectorConstruction.cpp]: If the `fRotation` was initiated correctly, then its size is always equal to 3, the three condition statements can be removed.

File: Core/DetectorConstruction.cpp [hunk 4]
```diff
@@ -1311,3 +1320,253 @@ void DetectorConstruction::CreateGeometryFile()
...
+                                                   dimensionsNew.at(2) * cm);
+          }
+
+          G4RotationMatrix* rotMatNew = new G4RotationMatrix();
+          if (phantomElements.at(idPhantElem.at(idToCombine)).fRotation.size() > 0) {
+            rotMatNew->rotateX(phantomElements.at(idPhantElem.at(idToCombine)).fRotation.at(0)*deg);
+            if (phantomElements.at(idPhantElem.at(idToCombine)).fRotation.size() > 1)
+              rotMatNew->rotateY(phantomElements.at(idPhantElem.at(idToCombine)).fRotation.at(1)*deg);
+            if (phantomElements.at(idPhantElem.at(idToCombine)).fRotation.size() > 2)
+              rotMatNew->rotateZ(phantomElements.at(idPhantElem.at(idToCombine)).fRotation.at(2)*deg);
```
Comment (kkacprzak) [inline Core/DetectorConstruction.cpp]: The rotation vector should always have the size == 3, why these checks are necessary?

File: Core/DetectorConstruction.cpp [hunk 5]
```diff
@@ -1311,3 +1320,253 @@ void DetectorConstruction::CreateGeometryFile()
...
+        if (elementMaterial == PhantomMaterial::aWater) {
+          G4Element* elH = new G4Element("Hydrogen", "H", z=1, 1.01*g/mole);
+          G4Element* elO = new G4Element("Oxygen", "O", z=8, 16*g/mole);
+          phantomElementMaterial = new G4Material("Water", phantomElements.at(i).fDensity*g/cm3, 2);
+          phantomElementMaterial->AddElement(elH, 2);
+          phantomElementMaterial->AddElement(elO, 1);
+          phantMatExt = new MaterialExtension(MaterialParameters::MaterialID::mWater,
+                                              "water" + std::to_string(i), phantomElementMaterial);
+        } else if (elementMaterial == PhantomMaterial::aPlastic) {
+          G4Element* elC = new G4Element("Carbon", "C", z=6., a=12.0107*g/mole);
```
Comment (kkacprzak) [inline Core/DetectorConstruction.cpp:1424]: Please avoid hard-coded values and use defined constants.

File: Core/PrimaryGenerator.cpp [hunk 6]
```diff
@@ -302,6 +314,43 @@ PrimaryGenerator::GetVerticesDistributionAlongStepVector(
...
+
+    if (mat) {
+      density = mat->GetDensity() / (g/cm3);
+      if (density < minDens)
+        density = minDens;
+    } else {
+      density = minDens;
+    }
+
+    reach = 19.378*std::pow(density, -1.53977); // test function of reach (in mm)
```
Comment (kkacprzak) [inline Core/PrimaryGenerator.cpp]: Please define the constants used in this function and comment on their values

File: Core/PrimaryGenerator.cpp [hunk 7]
```diff
@@ -514,16 +563,86 @@ void PrimaryGenerator::GenerateIsotope(SourceParams* sourceParams, G4Event* even
...
+  }
+
+  double DirectLF = nemaPoint.directLifetime;
+  G4double density = 0.1*g/cm3;
+
+  if (material)
+    density = material->GetDensity() / (g/cm3);
+
+  if (nemaPoint.directLFDensityDependent) {
+    DirectLF = 0.819151*exp(-0.517436*density);   //dummy function for test in ns
```
Comment (kkacprzak) [inline Core/PrimaryGenerator.cpp]: What are the constants in this function and can they be changed by the user?

File: Core/PrimaryGeneratorConstants.h [hunk 8]
```diff
@@ -23,6 +23,8 @@ namespace primary_generator_constants {
   // Cosmics muons parameters
   static const G4double MUON_CHARGE_RATIO = 1.2766;
   static const G4double MUON_MEAN_ENERGY_GEV = 4*GeV;
+  static const G4double DIRECT_PS_ANNIHILATION = 0.6;
```
Comment (kkacprzak) [inline Core/PrimaryGeneratorConstants.h:30]: Please add a short comment about these constants, where are they used and that they are not in the cosmic muon section

**Review by kkacprzak — CHANGES_REQUESTED** [review]
A great amount of job introducing the shapes, materials, generation modes etc.! All looks clear and straightforward. 

There is one point with the rotation vectors, that can be somewhat improved.
Also, please comment more on the usage of the probability functions for the positron path and consider moving the constants into the constants class.

**Comment by JPETTomographyTester** [comment]
Found 8 violations:

**Reporter**: cppcheck-raporter
**Rule**: unusedFunction
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/examples/basic/basic.cc L51

The function &#39;Construct&#39; is never used.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a-37406237>*

**Reporter**: cppcheck-raporter
**Rule**: redundantAssignment
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/examples/external/external.cc L99

Variable &#39;cone_mesh&#39; is reassigned a value before the old one has been used if variable is no semaphore variable.. Variable &#39;cone_mesh&#39; is reassigned a value before the old one has been used. Make sure that this variable is not used like a semaphore in a threading environment before simplifying this code.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a1979725627>*

**Reporter**: cppcheck-raporter
**Rule**: redundantAssignment
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/examples/external/external.cc L93

Variable &#39;cone_mesh&#39; is reassigned a value before the old one has been used if variable is no semaphore variable.. Variable &#39;cone_mesh&#39; is reassigned a value before the old one has been used. Make sure that this variable is not used like a semaphore in a threading environment before simplifying this code.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a1541790971>*

**Reporter**: cppcheck-raporter
**Rule**: unusedFunction
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/src/BuiltInReader.cc L81

The function &#39;BuiltIn&#39; is never used.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a229782319>*

**Reporter**: cppcheck-raporter
**Rule**: noExplicitConstructor
**Severity**: INFO
**File**: /var/jenkins_home/workspace/j-pet-geant-4-PRs-root-6/CADMesh/tests/Simulator.hh L29

Class &#39;DetectorConstruction&#39; has a constructor with 1 argument that is not explicit. Such constructors should in general be explicit for type safety reasons. Using the explicit keyword in the constructor means some mistakes when using the class can be avoided.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a-2004644777>*

**Reporter**: cppcheck-raporter
**Rule**: unreadVariable
**Severity**: INFO
**File**: Core/DetectorConstruction.cpp L1368

Variable &#39;rotationNew&#39; is assigned a value that is never used.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a1966920650>*

**Reporter**: cppcheck-raporter
**Rule**: variableScope
**Severity**: INFO
**File**: Core/DetectorConstruction.cpp L1332

The scope of the variable &#39;phantMatExt&#39; can be reduced. Warning: Be careful when fixing this message, especially when there are inner loops. Here is an example where cppcheck will write that the scope for &#39;i&#39; can be reduced:\012void f(int x)\012{\012    int i &#61; 0;\012    if (x) {\012        // it&#39;s safe to move &#39;int i &#61; 0;&#39; here\012        for (int n &#61; 0; n &lt; 10; ++n) {\012            // it is possible but not safe to move &#39;int i &#61; 0;&#39; here\012            do_something(&amp;i);\012        }\012    }\012}\012When you see this message it is always safe to reduce the variable scope 1 level.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a-778780786>*

**Reporter**: cppcheck-raporter
**Rule**: unmatchedSuppression
**Severity**: INFO
**File**: Core/DetectorConstruction.cpp L71

Unmatched suppression: unusedFunction

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a539693954>*

 *<ACCUMULATED-VIOLATIONS>*

**Review by kkacprzak — APPROVED** [review]

**Pull request merged**